### PR TITLE
Fix/restore and gate pox sunset

### PR DESCRIPTION
**Pull request by jcnelson**
This PR addresses #3343 by restoring the PoX sunset burn calculation, and gating its applicability based on the epoch.  The gating is confined to methods in `PoxConstants`; code that needs to verify whether or not a sunset is active uses these methods instead of directly reasoning about the current system epoch and block height (this "confines" the gates to one place).

In doing so, it's _possible_ that this PR also fixes #3281 by allowing calls to read-only functions in `.pox` when `.pox-2` is active.  The VM's special case handler had been too strict about this earlier.

### CODE REVIEW

**Comment by gregorycoppola** [comment]
Does this really fix #3341? That issue was closed.

**Comment by jcnelson** [comment]
Typo; updated to fix #3343.

File: testnet/stacks-node/src/tests/epoch_21.rs [hunk 1]
```diff
@@ -1245,3 +1250,301 @@ fn transition_adds_get_pox_addr_recipients() {
...
+    eprintln!("pox_info = {:?}", &pox_info);
+    assert_eq!(pox_info.current_cycle.is_pox_active, true);
+
+    // first full reward cycle with pox-2
+    assert_eq!(
+        &pox_info.contract_id,
+        &format!("ST000000000000000000002AMW42H.pox-2")
+    );
+
+    let burn_blocks = test_observer::get_burn_blocks();
```
Comment (gregorycoppola) [inline testnet/stacks-node/src/tests/epoch_21.rs:1494]: is there anything else to test about the pox amounts?

**Review by gregorycoppola — APPROVED** [review]
Overall, I think the tests look good.. i'm not totally sure if everything about the amounts of transfers is totally tested.

File: src/chainstate/burn/db/sortdb.rs [hunk 2]
```diff
@@ -268,7 +269,11 @@ impl FromRow<LeaderBlockCommitOp> for LeaderBlockCommitOp {
 
         let burn_fee = burn_fee_str
             .parse::<u64>()
-            .expect("DB Corruption: burn is not parseable as u64");
+            .expect("DB Corruption: Sunset burn is not parseable as u64");
```
Comment (obycode) [inline src/chainstate/burn/db/sortdb.rs:272]: This one should say "Burn fee is not parseable..."

File: testnet/stacks-node/src/tests/epoch_21.rs [hunk 3]
```diff
@@ -1245,3 +1250,301 @@ fn transition_adds_get_pox_addr_recipients() {
...
+    while sort_height <= epoch_21 + 1 {
+        next_block_and_wait(&mut btc_regtest_controller, &blocks_processed);
+        sort_height = channel.get_sortitions_processed();
+        eprintln!("Sort height pox-1: {} <= {}", sort_height, epoch_21);
+    }
+
+    let pox_info = get_pox_info(&http_origin);
+
+    // pox is still "active" despite unlock, because there's enough participation, and also even
+    // though the v1 blocok height has passed, the pox-2 contract won't be managing reward sets
```
Comment (obycode) [inline testnet/stacks-node/src/tests/epoch_21.rs:1432]:
```suggestion
    // though the v1 block height has passed, the pox-2 contract won't be managing reward sets
```

**Review by obycode — APPROVED** [review]
LGTM, just a few minor notes.

File: src/burnchains/burnchain.rs [hunk 4]
```diff
@@ -454,10 +474,63 @@ impl Burnchain {
         })
     }
 
+    /// BROKEN; DO NOT USE IN NEW CODE
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply src/burnchains/burnchain.rs]: Fixed

File: testnet/stacks-node/src/tests/epoch_21.rs (hunk 3 above)
Comment (jcnelson) [reply testnet/stacks-node/src/tests/epoch_21.rs]: Fixed

**Pull request merged**